File: csharp/src/Fory/MetaString.cs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

using System.Text;

namespace Apache.Fory;

public enum MetaStringEncoding : byte
{
    Utf8 = 0,
    LowerSpecial = 1,
    LowerUpperDigitSpecial = 2,
    FirstToLowerSpecial = 3,
    AllToLowerSpecial = 4,
}

public readonly struct MetaString : IEquatable<MetaString>
{
    private const int MaxMetaStringLength = 32_767;

    public MetaString(
        string value,
        MetaStringEncoding encoding,
        char specialChar1,
        char specialChar2,
        byte[] bytes)
    {
        if (value.Length >= MaxMetaStringLength)
        {
            throw new EncodingException("meta string too long");
        }

        if (encoding != MetaStringEncoding.Utf8 && bytes.Length == 0)
        {
            throw new EncodingException("encoded meta string cannot be empty");
        }

        Value = value;
        Encoding = encoding;
        SpecialChar1 = specialChar1;
        SpecialChar2 = specialChar2;
        Bytes = bytes;
        StripLastChar = encoding != MetaStringEncoding.Utf8 && (bytes[0] & 0x80) != 0;
    }

    public string Value { get; }

    public MetaStringEncoding Encoding { get; }

    public char SpecialChar1 { get; }

    public char SpecialChar2 { get; }

    public byte[] Bytes { get; }

    public bool StripLastChar { get; }

    public static MetaString Empty(char specialChar1, char specialChar2)
    {
        return new MetaString(string.Empty, MetaStringEncoding.Utf8, specialChar1, specialChar2, []);
    }

    public bool Equals(MetaString other)
    {
        return Value == other.Value &&
               Encoding == other.Encoding &&
               SpecialChar1 == other.SpecialChar1 &&
               SpecialChar2 == other.SpecialChar2 &&
               Bytes.AsSpan().SequenceEqual(other.Bytes);
    }

    public override bool Equals(object? obj)
    {
        return obj is MetaString other && Equals(other);
    }

    public override int GetHashCode()
    {
        HashCode hc = new();
        hc.Add(Value);
        hc.Add(Encoding);
        hc.Add(SpecialChar1);
        hc.Add(SpecialChar2);
        foreach (byte b in Bytes)
        {
            hc.Add(b);
        }

        return hc.ToHashCode();
    }
}

internal sealed class MetaStringEncoder
{
    private const int MaxMetaStringLength = 32_767;

    public MetaStringEncoder(char specialChar1, char specialChar2)
    {
        SpecialChar1 = specialChar1;
        SpecialChar2 = specialChar2;
    }

    public char SpecialChar1 { get; }

    public char SpecialChar2 { get; }

    public static MetaStringEncoder Namespace { get; } = new('.', '_');

    public static MetaStringEncoder TypeName { get; } = new('$', '_');

    public static MetaStringEncoder FieldName { get; } = new('$', '_');

    public MetaString Encode(string input)
    {
        return EncodeAuto(input, null);
    }

    public MetaString Encode(string input, IReadOnlyList<MetaStringEncoding> allowedEncodings)
    {
        return EncodeAuto(input, allowedEncodings);
    }

    public MetaString Encode(string input, MetaStringEncoding encoding)
    {
        if (input.Length >= MaxMetaStringLength)
        {
            throw new EncodingException("meta string too long");
        }

        if (input.Length == 0)
        {
            return MetaString.Empty(SpecialChar1, SpecialChar2);
        }

        if (encoding != MetaStringEncoding.Utf8 && !IsLatin(input))
        {
            throw new EncodingException("non-ASCII characters are not allowed for packed meta string");
        }

        return encoding switch
        {
            MetaStringEncoding.Utf8 => new MetaString(
                input,
                MetaStringEncoding.Utf8,
                SpecialChar1,
                SpecialChar2,
                Encoding.UTF8.GetBytes(input)),
            MetaStringEncoding.LowerSpecial => new MetaString(
                input,
                MetaStringEncoding.LowerSpecial,
                SpecialChar1,
                SpecialChar2,
                EncodeGeneric(input, 5, MapLowerSpecial)),
            MetaStringEncoding.LowerUpperDigitSpecial => new MetaString(
                input,
                MetaStringEncoding.LowerUpperDigitSpecial,
                SpecialChar1,
                SpecialChar2,
                EncodeGeneric(input, 6, MapLowerUpperDigitSpecial)),
            MetaStringEncoding.FirstToLowerSpecial => new MetaString(
                input,
                MetaStringEncoding.FirstToLowerSpecial,
                SpecialChar1,
                SpecialChar2,
                EncodeGeneric(LowerFirstAscii(input), 5, MapLowerSpecial)),
            MetaStringEncoding.AllToLowerSpecial => new MetaString(
                input,
                MetaStringEncoding.AllToLowerSpecial,
                SpecialChar1,
                SpecialChar2,
                EncodeGeneric(EscapeAllUpper(input), 5, MapLowerSpecial)),
            _ => throw new EncodingException($"unsupported meta string encoding: {encoding}"),
        };
    }

    private MetaString EncodeAuto(string input, IReadOnlyList<MetaStringEncoding>? allowedEncodings)
    {
        if (input.Length >= MaxMetaStringLength)
        {
            throw new EncodingException("meta string too long");
        }

        if (input.Length == 0)
        {
            return MetaString.Empty(SpecialChar1, SpecialChar2);
        }

        if (!IsLatin(input))
        {
            return new MetaString(input, MetaStringEncoding.Utf8, SpecialChar1, SpecialChar2, Encoding.UTF8.GetBytes(input));
        }

        MetaStringEncoding encoding = ChooseEncoding(input, allowedEncodings);
        return Encode(input, encoding);
    }

    private MetaStringEncoding ChooseEncoding(string input, IReadOnlyList<MetaStringEncoding>? allowedEncodings)
    {
        bool Allow(MetaStringEncoding encoding)
        {
            return allowedEncodings is null || allowedEncodings.Contains(encoding);
        }

        int digitCount = 0;
        int upperCount = 0;
        bool canLowerSpecial = true;
        bool canLowerUpperDigitSpecial = true;

        foreach (char c in input)
        {
            if (canLowerSpecial)
            {
                bool isValid = c is >= 'a' and <= 'z' || c is '.' or '_' or '$' or '|';
                if (!isValid)
                {
                    canLowerSpecial = false;
                }
            }

            if (canLowerUpperDigitSpecial)
            {
                bool isLower = c is >= 'a' and <= 'z';
                bool isUpper = c is >= 'A' and <= 'Z';
                bool isDigit = c is >= '0' and <= '9';
                bool isSpecial = c == SpecialChar1 || c == SpecialChar2;
                if (!(isLower || isUpper || isDigit || isSpecial))
                {
                    canLowerUpperDigitSpecial = false;
                }
            }

            if (c is >= '0' and <= '9')
            {
                digitCount++;
            }

            if (c is >= 'A' and <= 'Z')
            {
                upperCount++;
            }
        }

        if (canLowerSpecial && Allow(MetaStringEncoding.LowerSpecial))
        {
            return MetaStringEncoding.LowerSpecial;
        }

        if (canLowerUpperDigitSpecial)
        {
            if (digitCount != 0 && Allow(MetaStringEncoding.LowerUpperDigitSpecial))
            {
                return MetaStringEncoding.LowerUpperDigitSpecial;
            }

            if (upperCount == 1 &&
                char.IsUpper(input[0]) &&
                Allow(MetaStringEncoding.FirstToLowerSpecial))
            {
                return MetaStringEncoding.FirstToLowerSpecial;
            }

            if ((input.Length + upperCount) * 5 < input.Length * 6 && Allow(MetaStringEncoding.AllToLowerSpecial))
            {
                return MetaStringEncoding.AllToLowerSpecial;
            }

            if (Allow(MetaStringEncoding.LowerUpperDigitSpecial))
            {
                return MetaStringEncoding.LowerUpperDigitSpecial;
            }
        }

        return MetaStringEncoding.Utf8;
    }

    private byte[] EncodeGeneric(string input, int bitsPerChar, Func<char, byte> mapper)
    {
        int totalBits = input.Length * bitsPerChar + 1;
        int byteLength = (totalBits + 7) / 8;
        byte[] bytes = new byte[byteLength];
        int currentBit = 1;

        foreach (char c in input)
        {
            byte value = mapper(c);
            for (int i = bitsPerChar - 1; i >= 0; i--)
            {
                if (((value >> i) & 0x01) != 0)
                {
                    int bytePos = currentBit / 8;
                    int bitPos = currentBit % 8;
                    bytes[bytePos] |= (byte)(1 << (7 - bitPos));
                }

                currentBit++;
            }
        }

        if (byteLength * 8 >= totalBits + bitsPerChar)
        {
            bytes[0] |= 0x80;
        }

        return bytes;
    }

    private static byte MapLowerSpecial(char c)
    {
        if (c is >= 'a' and <= 'z')
        {
            return (byte)(c - 'a');
        }

        return c switch
        {
            '.' => 26,
            '_' => 27,
            '$' => 28,
            '|' => 29,
            _ => throw new EncodingException("unsupported character in LOWER_SPECIAL"),
        };
    }

    private byte MapLowerUpperDigitSpecial(char c)
    {
        if (c is >= 'a' and <= 'z')
        {
            return (byte)(c - 'a');
        }

        if (c is >= 'A' and <= 'Z')
        {
            return (byte)(26 + c - 'A');
        }

        if (c is >= '0' and <= '9')
        {
            return (byte)(52 + c - '0');
        }

        if (c == SpecialChar1)
        {
            return 62;
        }

        if (c == SpecialChar2)
        {
            return 63;
        }

        throw new EncodingException("unsupported character in LOWER_UPPER_DIGIT_SPECIAL");
    }

    private static string LowerFirstAscii(string input)
    {
        if (input.Length == 0)
        {
            return input;
        }

        return char.ToLowerInvariant(input[0]) + input[1..];
    }

    private static string EscapeAllUpper(string input)
    {
        StringBuilder sb = new(input.Length * 2);
        foreach (char c in input)
        {
            if (char.IsUpper(c))
            {
                sb.Append('|');
                sb.Append(char.ToLowerInvariant(c));
            }
            else
            {
                sb.Append(c);
            }
        }

        return sb.ToString();
    }

    private static bool IsLatin(string input)
    {
        foreach (char c in input)
        {
            if (c > 255)
            {
                return false;
            }
        }

        return true;
    }
}

internal sealed class MetaStringDecoder
{
    public MetaStringDecoder(char specialChar1, char specialChar2)
    {
        SpecialChar1 = specialChar1;
        SpecialChar2 = specialChar2;
    }

    public char SpecialChar1 { get; }

    public char SpecialChar2 { get; }

    public static MetaStringDecoder Namespace { get; } = new('.', '_');

    public static MetaStringDecoder TypeName { get; } = new('$', '_');

    public static MetaStringDecoder FieldName { get; } = new('$', '_');

    public MetaString Decode(byte[] bytes, MetaStringEncoding encoding)
    {
        string value = encoding switch
        {
            // C# intentionally preserves platform UTF-8 replacement behavior; Rust is the runtime
            // that provides checked UTF-8 string reads by default.
            MetaStringEncoding.Utf8 => Encoding.UTF8.GetString(bytes),
            MetaStringEncoding.LowerSpecial => DecodeGeneric(bytes, 5, UnmapLowerSpecial),
            MetaStringEncoding.LowerUpperDigitSpecial => DecodeGeneric(bytes, 6, UnmapLowerUpperDigitSpecial),
            MetaStringEncoding.FirstToLowerSpecial =>
                DecodeFirstToLowerSpecial(bytes),
            MetaStringEncoding.AllToLowerSpecial =>
                UnescapeAllUpper(DecodeGeneric(bytes, 5, UnmapLowerSpecial)),
            _ => throw new EncodingException($"unsupported meta string encoding: {encoding}"),
        };

        return new MetaString(value, encoding, SpecialChar1, SpecialChar2, bytes);
    }

    private string DecodeFirstToLowerSpecial(byte[] bytes)
    {
        string decoded = DecodeGeneric(bytes, 5, UnmapLowerSpecial);
        if (decoded.Length == 0)
        {
            return decoded;
        }

        return char.ToUpperInvariant(decoded[0]) + decoded[1..];
    }

    private string DecodeGeneric(byte[] bytes, int bitsPerChar, Func<byte, char> mapper)
    {
        if (bytes.Length == 0)
        {
            return string.Empty;
        }

        bool stripLast = (bytes[0] & 0x80) != 0;
        int totalBits = bytes.Length * 8;
        int bitIndex = 1;
        StringBuilder sb = new(bytes.Length);
        while (bitIndex + bitsPerChar <= totalBits &&
               !(stripLast && (bitIndex + 2 * bitsPerChar > totalBits)))
        {
            byte value = 0;
            for (var i = 0; i < bitsPerChar; i++)
            {
                int byteIndex = bitIndex / 8;
                int intra = bitIndex % 8;
                byte bit = (byte)((bytes[byteIndex] >> (7 - intra)) & 0x01);
                value = (byte)((value << 1) | bit);
                bitIndex++;
            }

            sb.Append(mapper(value));
        }

        return sb.ToString();
    }

    private static char UnmapLowerSpecial(byte value)
    {
        return value switch
        {
            <= 25 => (char)('a' + value),
            26 => '.',
            27 => '_',
            28 => '$',
            29 => '|',
            _ => throw new EncodingException("invalid LOWER_SPECIAL value"),
        };
    }

    private char UnmapLowerUpperDigitSpecial(byte value)
    {
        return value switch
        {
            <= 25 => (char)('a' + value),
            <= 51 => (char)('A' + value - 26),
            <= 61 => (char)('0' + value - 52),
            62 => SpecialChar1,
            63 => SpecialChar2,
            _ => throw new EncodingException("invalid LOWER_UPPER_DIGIT_SPECIAL value"),
        };
    }

    private static string UnescapeAllUpper(string input)
    {
        StringBuilder sb = new(input.Length);
        for (int i = 0; i < input.Length; i++)
        {
            char c = input[i];
            if (c == '|' && i + 1 < input.Length)
            {
                i++;
                sb.Append(char.ToUpperInvariant(input[i]));
            }
            else
            {
                sb.Append(c);
            }
        }

        return sb.ToString();
    }
}

File: dart/packages/fory/lib/src/meta/meta_string.dart
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import 'dart:convert';
import 'dart:typed_data';

import 'package:fory/src/types/int64.dart';
import 'package:fory/src/util/hash_util.dart';

const int metaStringUtf8Encoding = 0;
const int metaStringLowerSpecialEncoding = 1;
const int metaStringLowerUpperDigitSpecialEncoding = 2;
const int metaStringFirstToLowerSpecialEncoding = 3;
const int metaStringAllToLowerSpecialEncoding = 4;

const int metaStringSmallThreshold = 16;
const int typeDefSmallFieldCountThreshold = 0x1f;
const int typeDefRegisterByNameFlag = 0x20;
const int typeDefCompatibleFlag = 0x40;
const int typeDefStructFlag = 0x80;
const int typeDefBigFieldNameThreshold = 0x0f;
const int typeDefBigNameThreshold = 0x3f;

const List<int> _packageNameCompactEncodings = <int>[
  metaStringUtf8Encoding,
  metaStringAllToLowerSpecialEncoding,
  metaStringLowerUpperDigitSpecialEncoding,
];
const List<int> _typeNameCompactEncodings = <int>[
  metaStringUtf8Encoding,
  metaStringAllToLowerSpecialEncoding,
  metaStringLowerUpperDigitSpecialEncoding,
  metaStringFirstToLowerSpecialEncoding,
];
const List<int> _fieldNameCompactEncodings = <int>[
  metaStringUtf8Encoding,
  metaStringAllToLowerSpecialEncoding,
  metaStringLowerUpperDigitSpecialEncoding,
];

final class EncodedMetaString {
  final Uint8List bytes;
  final int encoding;
  final Int64 hash;
  final int firstWord0;
  final int firstWord1;
  final int secondWord0;
  final int secondWord1;

  static final EncodedMetaString empty =
      EncodedMetaString(Uint8List(0), metaStringUtf8Encoding);

  EncodedMetaString(this.bytes, this.encoding)
      : hash = metaStringHash(bytes, encoding: encoding),
        firstWord0 = _packLittleEndian32(bytes, 0),
        firstWord1 = _packLittleEndian32(bytes, 4),
        secondWord0 = _packLittleEndian32(bytes, 8),
        secondWord1 = _packLittleEndian32(bytes, 12);

  int get length => bytes.length;

  bool get isSmall => bytes.length <= metaStringSmallThreshold;

  bool matchesPacked(
    int packedEncoding,
    int packedLength,
    int packedWord0,
    int packedWord1,
    int packedWord2,
    int packedWord3,
  ) {
    return encoding == packedEncoding &&
        bytes.length == packedLength &&
        firstWord0 == packedWord0 &&
        firstWord1 == packedWord1 &&
        secondWord0 == packedWord2 &&
        secondWord1 == packedWord3;
  }
}

EncodedMetaString encodePackageMetaString(String value) => _encodeMetaString(
      value,
      specialChar1: '.',
      specialChar2: '_',
      allowedEncodings: _packageNameCompactEncodings,
    );

EncodedMetaString encodeTypeNameMetaString(String value) => _encodeMetaString(
      value,
      specialChar1: r'$',
      specialChar2: '_',
      allowedEncodings: _typeNameCompactEncodings,
    );

EncodedMetaString encodeFieldNameMetaString(String value) => _encodeMetaString(
      value,
      specialChar1: r'$',
      specialChar2: '_',
      allowedEncodings: _fieldNameCompactEncodings,
    );

int packageNameCompactEncoding(int encoding) =>
    _compactEncodingIndex(encoding, _packageNameCompactEncodings, 'package');

int typeNameCompactEncoding(int encoding) =>
    _compactEncodingIndex(encoding, _typeNameCompactEncodings, 'type name');

int fieldNameCompactEncoding(int encoding) =>
    _compactEncodingIndex(encoding, _fieldNameCompactEncodings, 'field name');

int packageNameEncoding(int compactEncoding) => _compactEncodingValue(
      compactEncoding,
      _packageNameCompactEncodings,
      'package',
    );

int typeNameEncoding(int compactEncoding) => _compactEncodingValue(
      compactEncoding,
      _typeNameCompactEncodings,
      'type name',
    );

String decodePackageMetaString(List<int> bytes, int encoding) =>
    decodeMetaString(
      bytes,
      encoding,
      specialChar1: '.',
      specialChar2: '_',
    );

String decodeTypeNameMetaString(List<int> bytes, int encoding) =>
    decodeMetaString(
      bytes,
      encoding,
      specialChar1: r'$',
      specialChar2: '_',
    );

String decodePackageName(List<int> bytes, int compactEncoding) =>
    decodePackageMetaString(
      bytes,
      _compactEncodingValue(
        compactEncoding,
        _packageNameCompactEncodings,
        'package',
      ),
    );

String decodeTypeName(List<int> bytes, int compactEncoding) =>
    decodeTypeNameMetaString(
      bytes,
      _compactEncodingValue(
        compactEncoding,
        _typeNameCompactEncodings,
        'type name',
      ),
    );

String decodeFieldName(List<int> bytes, int compactEncoding) =>
    decodeMetaString(
      bytes,
      _compactEncodingValue(
        compactEncoding,
        _fieldNameCompactEncodings,
        'field name',
      ),
      specialChar1: r'$',
      specialChar2: '_',
    );

String decodeMetaString(
  List<int> bytes,
  int encoding, {
  required String specialChar1,
  required String specialChar2,
}) {
  if (bytes.isEmpty) {
    return '';
  }
  switch (encoding) {
    case metaStringUtf8Encoding:
      return utf8.decode(bytes);
    case metaStringLowerSpecialEncoding:
      return _decodeLowerSpecial(bytes);
    case metaStringLowerUpperDigitSpecialEncoding:
      return _decodeLowerUpperDigitSpecial(
        bytes,
        specialChar1: specialChar1,
        specialChar2: specialChar2,
      );
    case metaStringFirstToLowerSpecialEncoding:
      final value = _decodeLowerSpecial(bytes);
      if (value.isEmpty) {
        return value;
      }
      return '${value[0].toUpperCase()}${value.substring(1)}';
    case metaStringAllToLowerSpecialEncoding:
      return _decodeAllToLowerSpecial(bytes);
    default:
      throw StateError('Unsupported meta-string encoding $encoding.');
  }
}

int _packLittleEndian32(List<int> bytes, int offset) {
  if (offset >= bytes.length) {
    return 0;
  }
  final end = offset + 4;
  var value = 0;
  for (var index = offset; index < bytes.length && index < end; index += 1) {
    value |= (bytes[index] & 0xff) << ((index - offset) * 8);
  }
  return value;
}

EncodedMetaString _encodeMetaString(
  String value, {
  required String specialChar1,
  required String specialChar2,
  required List<int> allowedEncodings,
}) {
  if (value.isEmpty) {
    return EncodedMetaString.empty;
  }
  for (final codeUnit in value.codeUnits) {
    if (codeUnit > 0x7f) {
      return EncodedMetaString(
        Uint8List.fromList(utf8.encode(value)),
        metaStringUtf8Encoding,
      );
    }
  }
  final encoding = _chooseEncoding(
    value,
    specialChar1: specialChar1,
    specialChar2: specialChar2,
    allowedEncodings: allowedEncodings,
  );
  switch (encoding) {
    case metaStringLowerSpecialEncoding:
      return EncodedMetaString(
        _encodeLowerSpecial(value.codeUnits),
        encoding,
      );
    case metaStringLowerUpperDigitSpecialEncoding:
      return EncodedMetaString(
        _encodeLowerUpperDigitSpecial(
          value.codeUnits,
          specialChar1: specialChar1,
          specialChar2: specialChar2,
        ),
        encoding,
      );
    case metaStringFirstToLowerSpecialEncoding:
      final codeUnits = value.codeUnits.toList(growable: false);
      codeUnits[0] = _toLowerAscii(codeUnits.first);
      return EncodedMetaString(
        _encodeLowerSpecial(codeUnits),
        encoding,
      );
    case metaStringAllToLowerSpecialEncoding:
      return EncodedMetaString(
        _encodeAllToLowerSpecial(value.codeUnits),
        encoding,
      );
    case metaStringUtf8Encoding:
      return EncodedMetaString(
        Uint8List.fromList(utf8.encode(value)),
        encoding,
      );
    default:
      throw StateError('Unsupported meta-string encoding $encoding.');
  }
}

int _chooseEncoding(
  String value, {
  required String specialChar1,
  required String specialChar2,
  required List<int> allowedEncodings,
}) {
  var canLowerUpperDigitSpecial = true;
  var canLowerSpecial = true;
  var digitCount = 0;
  var upperCount = 0;
  final codeUnits = value.codeUnits;
  for (final codeUnit in codeUnits) {
    if (canLowerUpperDigitSpecial &&
        !_isLowerUpperDigitSpecialChar(
          codeUnit,
          specialChar1: specialChar1.codeUnitAt(0),
          specialChar2: specialChar2.codeUnitAt(0),
        )) {
      canLowerUpperDigitSpecial = false;
    }
    if (canLowerSpecial && !_isLowerSpecialChar(codeUnit)) {
      canLowerSpecial = false;
    }
    if (_isAsciiDigit(codeUnit)) {
      digitCount += 1;
    }
    if (_isUpperAscii(codeUnit)) {
      upperCount += 1;
    }
  }
  if (canLowerSpecial &&
      allowedEncodings.contains(metaStringLowerSpecialEncoding)) {
    return metaStringLowerSpecialEncoding;
  }
  if (canLowerUpperDigitSpecial) {
    if (digitCount != 0 &&
        allowedEncodings.contains(metaStringLowerUpperDigitSpecialEncoding)) {
      return metaStringLowerUpperDigitSpecialEncoding;
    }
    if (upperCount == 1 &&
        _isUpperAscii(codeUnits.first) &&
        allowedEncodings.contains(metaStringFirstToLowerSpecialEncoding)) {
      return metaStringFirstToLowerSpecialEncoding;
    }
    if ((codeUnits.length + upperCount) * 5 < codeUnits.length * 6 &&
        allowedEncodings.contains(metaStringAllToLowerSpecialEncoding)) {
      return metaStringAllToLowerSpecialEncoding;
    }
    if (allowedEncodings.contains(metaStringLowerUpperDigitSpecialEncoding)) {
      return metaStringLowerUpperDigitSpecialEncoding;
    }
  }
  return metaStringUtf8Encoding;
}

Uint8List _encodeLowerSpecial(List<int> codeUnits) =>
    _encodeGeneric(codeUnits, 5, _encodeLowerSpecialChar);

Uint8List _encodeLowerUpperDigitSpecial(
  List<int> codeUnits, {
  required String specialChar1,
  required String specialChar2,
}) {
  return _encodeGeneric(
    codeUnits,
    6,
    (codeUnit) => _encodeLowerUpperDigitSpecialChar(
      codeUnit,
      specialChar1: specialChar1.codeUnitAt(0),
      specialChar2: specialChar2.codeUnitAt(0),
    ),
  );
}

Uint8List _encodeAllToLowerSpecial(List<int> codeUnits) {
  final escaped = <int>[];
  for (final codeUnit in codeUnits) {
    if (_isUpperAscii(codeUnit)) {
      escaped.add('|'.codeUnitAt(0));
      escaped.add(_toLowerAscii(codeUnit));
    } else {
      escaped.add(codeUnit);
    }
  }
  return _encodeLowerSpecial(escaped);
}

Uint8List _encodeGeneric(
  List<int> codeUnits,
  int bitsPerChar,
  int Function(int codeUnit) encodeChar,
) {
  final totalBits = codeUnits.length * bitsPerChar + 1;
  final byteLength = (totalBits + 7) >> 3;
  final bytes = Uint8List(byteLength);
  var byteIndex = 0;
  var bitIndex = 1;
  var charIndex = 0;
  var charBitRemain = bitsPerChar;
  while (charIndex < codeUnits.length) {
    final charValue = encodeChar(codeUnits[charIndex]);
    final nowByteRemain = 8 - bitIndex;
    if (nowByteRemain >= charBitRemain) {
      final mask = (1 << charBitRemain) - 1;
      bytes[byteIndex] |= (charValue & mask) << (nowByteRemain - charBitRemain);
      bitIndex += charBitRemain;
      if (bitIndex == 8) {
        byteIndex += 1;
        bitIndex = 0;
      }
      charIndex += 1;
      charBitRemain = bitsPerChar;
      continue;
    }
    final mask = (1 << nowByteRemain) - 1;
    bytes[byteIndex] |= (charValue >> (charBitRemain - nowByteRemain)) & mask;
    byteIndex += 1;
    bitIndex = 0;
    charBitRemain -= nowByteRemain;
  }
  final stripLastChar = bytes.length * 8 >= totalBits + bitsPerChar;
  if (stripLastChar) {
    bytes[0] |= 0x80;
  }
  return bytes;
}

String _decodeLowerSpecial(List<int> bytes) {
  final buffer = StringBuffer();
  final totalBits = bytes.length * 8;
  final stripLastChar = (bytes[0] & 0x80) != 0;
  var bitIndex = 1;
  while (bitIndex + 5 <= totalBits &&
      !(stripLastChar && (bitIndex + 10 > totalBits))) {
    final value = _readPackedValue(bytes, bitIndex, 5);
    bitIndex += 5;
    buffer.writeCharCode(_decodeLowerSpecialChar(value));
  }
  return buffer.toString();
}

String _decodeLowerUpperDigitSpecial(
  List<int> bytes, {
  required String specialChar1,
  required String specialChar2,
}) {
  final buffer = StringBuffer();
  final totalBits = bytes.length * 8;
  final stripLastChar = (bytes[0] & 0x80) != 0;
  var bitIndex = 1;
  while (bitIndex + 6 <= totalBits &&
      !(stripLastChar && (bitIndex + 12 > totalBits))) {
    final value = _readPackedValue(bytes, bitIndex, 6);
    bitIndex += 6;
    buffer.writeCharCode(
      _decodeLowerUpperDigitSpecialChar(
        value,
        specialChar1: specialChar1,
        specialChar2: specialChar2,
      ),
    );
  }
  return buffer.toString();
}

String _decodeAllToLowerSpecial(List<int> bytes) {
  final decoded = _decodeLowerSpecial(bytes);
  final buffer = StringBuffer();
  for (var index = 0; index < decoded.length; index += 1) {
    final character = decoded[index];
    if (character == '|' && index + 1 < decoded.length) {
      index += 1;
      buffer.write(decoded[index].toUpperCase());
      continue;
    }
    buffer.write(character);
  }
  return buffer.toString();
}

int _readPackedValue(List<int> bytes, int bitIndex, int width) {
  var value = 0;
  for (var offset = 0; offset < width; offset += 1) {
    final absoluteBit = bitIndex + offset;
    final byteIndex = absoluteBit ~/ 8;
    final intraByteIndex = 7 - (absoluteBit % 8);
    value = (value << 1) | ((bytes[byteIndex] >> intraByteIndex) & 1);
  }
  return value;
}

int _encodeLowerSpecialChar(int codeUnit) {
  if (codeUnit >= 0x61 && codeUnit <= 0x7a) {
    return codeUnit - 0x61;
  }
  switch (codeUnit) {
    case 0x2e:
      return 26;
    case 0x5f:
      return 27;
    case 0x24:
      return 28;
    case 0x7c:
      return 29;
    default:
      throw StateError(
        'Unsupported character for LOWER_SPECIAL encoding: ${String.fromCharCode(codeUnit)}.',
      );
  }
}

int _decodeLowerSpecialChar(int value) {
  if (value >= 0 && value <= 25) {
    return 'a'.codeUnitAt(0) + value;
  }
  switch (value) {
    case 26:
      return '.'.codeUnitAt(0);
    case 27:
      return '_'.codeUnitAt(0);
    case 28:
      return r'$'.codeUnitAt(0);
    case 29:
      return '|'.codeUnitAt(0);
    default:
      throw StateError('Invalid LOWER_SPECIAL value $value.');
  }
}

int _encodeLowerUpperDigitSpecialChar(
  int codeUnit, {
  required int specialChar1,
  required int specialChar2,
}) {
  if (codeUnit >= 0x61 && codeUnit <= 0x7a) {
    return codeUnit - 0x61;
  }
  if (codeUnit >= 0x41 && codeUnit <= 0x5a) {
    return 26 + (codeUnit - 0x41);
  }
  if (codeUnit >= 0x30 && codeUnit <= 0x39) {
    return 52 + (codeUnit - 0x30);
  }
  if (codeUnit == specialChar1) {
    return 62;
  }
  if (codeUnit == specialChar2) {
    return 63;
  }
  throw StateError(
    'Unsupported character for LOWER_UPPER_DIGIT_SPECIAL encoding: ${String.fromCharCode(codeUnit)}.',
  );
}

int _decodeLowerUpperDigitSpecialChar(
  int value, {
  required String specialChar1,
  required String specialChar2,
}) {
  if (value >= 0 && value <= 25) {
    return 'a'.codeUnitAt(0) + value;
  }
  if (value >= 26 && value <= 51) {
    return 'A'.codeUnitAt(0) + (value - 26);
  }
  if (value >= 52 && value <= 61) {
    return '0'.codeUnitAt(0) + (value - 52);
  }
  if (value == 62) {
    return specialChar1.codeUnitAt(0);
  }
  if (value == 63) {
    return specialChar2.codeUnitAt(0);
  }
  throw StateError('Invalid LOWER_UPPER_DIGIT_SPECIAL value $value.');
}

int _compactEncodingIndex(
  int encoding,
  List<int> encodings,
  String label,
) {
  final index = encodings.indexOf(encoding);
  if (index == -1) {
    throw StateError('Unsupported $label encoding $encoding.');
  }
  return index;
}

int _compactEncodingValue(
  int compactEncoding,
  List<int> encodings,
  String label,
) {
  if (compactEncoding < 0 || compactEncoding >= encodings.length) {
    throw StateError('Unsupported $label compact encoding $compactEncoding.');
  }
  return encodings[compactEncoding];
}

bool _isLowerUpperDigitSpecialChar(
  int codeUnit, {
  required int specialChar1,
  required int specialChar2,
}) {
  return (codeUnit >= 0x61 && codeUnit <= 0x7a) ||
      (codeUnit >= 0x41 && codeUnit <= 0x5a) ||
      (codeUnit >= 0x30 && codeUnit <= 0x39) ||
      codeUnit == specialChar1 ||
      codeUnit == specialChar2;
}

bool _isLowerSpecialChar(int codeUnit) {
  return (codeUnit >= 0x61 && codeUnit <= 0x7a) ||
      codeUnit == 0x2e ||
      codeUnit == 0x5f ||
      codeUnit == 0x24 ||
      codeUnit == 0x7c;
}

bool _isUpperAscii(int codeUnit) => codeUnit >= 0x41 && codeUnit <= 0x5a;

bool _isAsciiDigit(int codeUnit) => codeUnit >= 0x30 && codeUnit <= 0x39;

int _toLowerAscii(int codeUnit) =>
    _isUpperAscii(codeUnit) ? codeUnit + 0x20 : codeUnit;
